### PR TITLE
[Algorithm] Discrete CQL

### PR DESCRIPTION
**Pull request by BY571**
## Description

Adds discrete (DQN) CQL objective and example

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation (update in the documentation)
- [X] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [X] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by BY571** [comment]
image


Converges on Cartpole as expected. Just needs some cleanup + tests

File: examples/cql/discrete_cql_online.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        cfg.collector.env_per_collector
+        * cfg.collector.frames_per_batch
+        * cfg.optim.utd_ratio
+    )
+    prb = cfg.replay_buffer.prb
+    eval_rollout_steps = cfg.env.max_episode_steps
+    eval_iter = cfg.logger.eval_iter
+    frames_per_batch = cfg.collector.frames_per_batch
+
+    sampling_start = time.time()
```
Comment (vmoens) [inline examples/cql/discrete_cql_online.py:95]:
```suggestion
    collected_frames = 0
    pbar = tqdm.tqdm(total=cfg.collector.total_frames)

    init_random_frames = cfg.collector.init_random_frames
    num_updates = int(
        cfg.collector.env_per_collector
        * cfg.collector.frames_per_batch
        * cfg.optim.utd_ratio
    )
    prb = cfg.replay_buffer.prb
    eval_rollout_steps = cfg.env.max_episode_steps
    eval_iter = cfg.logger.eval_iter
    frames_per_batch = cfg.collector.frames_per_batch

    start_time = sampling_start = time.time()
```

File: test/test_cost.py [hunk 2]
```diff
@@ -5164,6 +5165,367 @@ def test_cql_batcher(
...
+    ):
+        # Actor
+        if action_spec_type == "one_hot":
+            action_spec = OneHotDiscreteTensorSpec(action_dim)
+        elif action_spec_type == "categorical":
+            action_spec = DiscreteTensorSpec(action_dim)
+        # elif action_spec_type == "nd_bounded":
+        #     action_spec = BoundedTensorSpec(
+        #         -torch.ones(action_dim), torch.ones(action_dim), (action_dim,)
+        #     )
```
Comment (vmoens) [inline test/test_cost.py:5189]:
```suggestion
```

File: test/test_cost.py [hunk 3]
```diff
@@ -5164,6 +5165,367 @@ def test_cql_batcher(
...
+        else:
+            raise ValueError(f"Wrong {action_spec_type}")
```
Comment (vmoens) [inline test/test_cost.py:5191]:
```suggestion
            raise ValueError(f"Wrong action spec type: {action_spec_type}")
```

File: torchrl/objectives/cql.py [hunk 4]
```diff
@@ -793,3 +797,330 @@ def _alpha(self):
...
+            priority (NestedKey): The input tensordict key where the target priority is written to.
+                Defaults to ``"td_error"``.
+            reward (NestedKey): The input tensordict key where the reward is expected.
+                Will be used for the underlying value estimator. Defaults to ``"reward"``.
+            done (NestedKey): The key in the input TensorDict that indicates
+                whether a trajectory is done. Will be used for the underlying value estimator.
+                Defaults to ``"done"``.
+            terminated (NestedKey): The key in the input TensorDict that indicates
+                whether a trajectory is terminated. Will be used for the underlying value estimator.
+                Defaults to ``"terminated"``.
```
Comment (vmoens) [inline torchrl/objectives/cql.py:904]: this does not match the keys below

File: torchrl/objectives/cql.py [hunk 5]
```diff
@@ -793,3 +797,330 @@ def _alpha(self):
...
+                # let's try with action_space then
+                try:
+                    action_space = value_network.action_space
+                except AttributeError:
+                    raise ValueError(self.ACTION_SPEC_ERROR)
+        if action_space is None:
+            warnings.warn(
+                "action_space was not specified. DiscreteCQLLoss will default to 'one-hot'."
+                "This behaviour will be deprecated soon and a space will have to be passed."
+                "Check the DiscreteCQLLoss documentation to see how to pass the action space. "
```
Comment (vmoens) [inline torchrl/objectives/cql.py:966]:
```suggestion
                "action_space was not specified. DiscreteCQLLoss will default to 'one-hot'. "
                "This behaviour will be deprecated soon and a space will have to be passed. "
                "Check the DiscreteCQLLoss documentation to see how to pass the action space."
```

File: torchrl/objectives/cql.py [hunk 6]
```diff
@@ -793,3 +797,330 @@ def _alpha(self):
...
+    @_cache_values
+    def _cached_detached_target_value_params(self):
+        return self.target_value_network_params.detach()
+
+    def cql_loss(self, q_values, current_action):
+        """Computes the CQL loss for a batch of Q-values and one-hot encoded actions."""
+        logsumexp = torch.logsumexp(q_values, dim=-1, keepdim=True)
+        q_a = (q_values * current_action).sum(dim=-1, keepdim=True)
+
+        return (logsumexp - q_a).mean()
```
Comment (vmoens) [inline torchrl/objectives/cql.py:1126]: can we return metadata too, like we're hoping to do for all losses in the future?

File: torchrl/objectives/cql.py [hunk 7]
```diff
@@ -793,3 +797,330 @@ def _alpha(self):
...
+        if self._in_keys is None:
+            self._set_in_keys()
+        return self._in_keys
+
+    @in_keys.setter
+    def in_keys(self, values):
+        self._in_keys = values
+
+    @dispatch
+    def forward(self, tensordict: TensorDictBase) -> TensorDict:
```
Comment (vmoens) [inline torchrl/objectives/cql.py:1057]: this should just be a couple of lines with dqn_loss and cql_loss IMO

**Review by vmoens — COMMENTED** [review]
Great work! I left some high level comments, can you have a look?
Thanks for this

File: torchrl/objectives/cql.py (hunk 7 above)
Comment (BY571) [reply torchrl/objectives/cql.py:1057]: I actually tried to inherit from the DQN class and then do something like super.forward(tensordict) and only have the cql_loss calculation added but I got circular importing issues. Do you have any suggestions?

Comment (vmoens) [reply torchrl/objectives/cql.py:1057]: Oh I wasn't suggesting to inherit from DQN, it's ok if they're separated. But the forward should just be a composition of `loss_actor` and `loss_critic` like we did in other losses (eg, TD3), where each sub-loss returns a tensor and a dict of metadata.

Comment (BY571) [reply torchrl/objectives/cql.py]: ah, got it! Should be adapted accordingly now.

**Review by vmoens — COMMENTED** [review]
I don't understand it.
CQL loss is called in value loss, not in forward, why is that?
Why do we call item() on CQL loss value? Conventionally all losses in the output tensordict of a loss module should be differentials. 
Can you give me some context?

**Comment by BY571** [comment]
> I don't understand it. CQL loss is called in value loss, not in forward, why is that? Why do we call item() on CQL loss value? Conventionally all losses in the output tensordict of a loss module should be differentials. Can you give me some context?

The CQL loss is more like an auxiliary term for the value loss not for a separate model like the actor. It just augments the value loss.  We could separate it but then we would need to forward pass through the model again to obtain the current q values, which would slow down the process and I think there is no need to obtain only the cql loss as in itself it's incomplete.

File: torchrl/objectives/cql.py [hunk 8]
```diff
@@ -793,3 +797,355 @@ def _alpha(self):
...
+            inplace=True,
+        )
+        loss = distance_loss(pred_val_index, target_value, self.loss_function).mean()
+
+        metadata = {
+            "td_error": td_error.mean(0).detach(),
+            "loss_cql": cql_loss.item(),
+            "pred_value": pred_val.mean().detach(),
+            "target_value": target_value.mean().detach(),
+        }
```
[the remark this answers is not in the document]
Comment (BY571) [reply torchrl/objectives/cql.py]: you are right, I must have deleted it. Sorry for the confusion, I just updated and fixed it :)

**Comment by vmoens** [comment]
What do you think of https://github.com/BY571/rl/pull/1? I think being able to run ablation studies has some value.

We need to fix the categorical case.

**Comment by BY571** [comment]
> What do you think of [BY571#1](https://github.com/BY571/rl/pull/1)? I think being able to run ablation studies has some value.
> 
> We need to fix the categorical case.

I think yes, if someone wants to check how the cql loss term influences the agent performance and want to have simple "on/off" capability it makes sense. The changes you did look good, I also pushed some adaption for the categorical case to calculate the cql loss.

**Comment by vmoens** [comment]
Cool LMK when you've merge the PR

File: torchrl/objectives/cql.py [hunk 9]
```diff
@@ -793,3 +797,359 @@ def _alpha(self):
...
+        self._value_estimator.set_keys(**tensor_keys)
+
+    @property
+    def device(self) -> torch.device:
+        for p in self.parameters():
+            return p.device
+        raise RuntimeError(
+            "At least one of the networks of CQLLoss must have trainable " "parameters."
+        )
```
Comment (vmoens) [inline torchrl/objectives/cql.py:1047]: loss.device will be deprecated in the future to allow losses to be distributed on multiple devices. Let's not add it to new losses

File: torchrl/objectives/cql.py [hunk 10]
```diff
@@ -793,3 +797,359 @@ def _alpha(self):
...
+
+    def __init__(
+        self,
+        value_network: Union[QValueActor, nn.Module],
+        *,
+        loss_function: Optional[str] = "l2",
+        delay_value: bool = True,
+        gamma: float = None,
+        action_space=None,
+        priority_key: str = None,
```
Comment (vmoens) [inline torchrl/objectives/cql.py:934]: passing keys to constructors will be deprecated for existing losses, let's not add it to new ones.

**Comment by BY571** [comment]
Just merged and fixed the open issues. Let me know what you think.
Also, thank you for insisting on making the losses separate, I took advantage of it already and compared base DQN vs DQN+CQL loss :)
image

**Comment by vmoens** [comment]
That looks great!
There are still 19 broken tests in the new test class and the example isn't running either.

**Review by vmoens — APPROVED** [review]
Cool let's merge this!
